src/hooks/filters.ts

```diff
@@ -0,0 +1,294 @@
+ * @returns filteredQueries - the filtered queries
+ * @returns checkedChange - update the filters based on the checked state of a checkbox
+ * @returns reset - reset the filters to their initial state
+ */
+export function useFilters(queries: OracleQueryUI[]) {
+  const initialState: State = {
+    filters: {
+      project: { All: { checked: true, count: 0 } },
+      chainName: { All: { checked: true, count: 0 } },
+      oracleType: { All: { checked: true, count: 0 } },
+    },
+    checkedFilters: {
+      project: [],
+      chainName: [],
+      oracleType: [],
+    },
+    filteredQueries: cloneDeep(queries),
+  };
+  const [state, dispatch] = useReducer(reducer, initialState);
+
+  useEffect(() => {
+    dispatch({ type: "make-entries" });
+  }, [queries]);
+
+  function reducer(state: State, action: Action) {
+    switch (action.type) {
+      case "make-entries": {
+        const newState = cloneDeep(state);
+
+        queries.forEach((query) => {
+          const { project, chainName, oracleType } = query;
+
+          // for all filters, increment the count for the "All" item
+          newState.filters.project.All.count++;
+          newState.filters.chainName.All.count++;
+          newState.filters.oracleType.All.count++;
+
+          if (newState.filters.project[project]) {
+            newState.filters.project[project].count++;
+          } else {
+            newState.filters.project[project] = { checked: false, count: 1 };
+          }
+
+          if (newState.filters.chainName[chainName]) {
+            newState.filters.chainName[chainName].count++;
+          } else {
+            newState.filters.chainName[chainName] = {
+              checked: false,
+              count: 1,
+            };
+          }
+
+          if (newState.filters.oracleType[oracleType]) {
+            newState.filters.oracleType[oracleType].count++;
+          } else {
+            newState.filters.oracleType[oracleType] = {
+              checked: false,
+              count: 1,
+            };
+          }
+        });
+
+        newState.checkedFilters = makeCheckedFilters(newState);
+        newState.filteredQueries = makeFilteredQueries(newState);
+
+        return newState;
+      }
+      case "checked-change": {
+        const newState = cloneDeep(state);
+        const { filters } = newState;
+        const { filterName, checked, itemName } = action.payload;
+        const items = filters[filterName];
+        const newFilters = makeFilters({
+          items,
+          checked,
+          itemName,
+        });
+
+        newState.filters[filterName] = newFilters;
+        newState.checkedFilters = makeCheckedFilters(newState);
+        newState.filteredQueries = makeFilteredQueries(newState);
+
+        return newState;
+      }
+      case "reset": {
+        const newState = cloneDeep(state);
+
+        Object.values(newState.filters).forEach((filter) => {
```

Comment:
just a nit, but it seems like you could pre-process the filters at the top of the file and then remove all this logic from the reducer, and just clone the processed filters

Reply:
I don't think this would work because the items in each of the filters are generated from the queries. I didn't extract this to a function because here is the only place this logic is used